### PR TITLE
Implements Py3-compatible rich comparison for all Bio.PDB objects

### PR DESCRIPTION
**Pull request by JoaoRodrigues**
This pull request addresses issue #1356 and refactors #1189 and PR #1291. It implements rich comparisons for every object of the SMCRA hierarchy, including disordered wrappers. This allows direct comparison of any two objects and allows calling the `sorted` method directly, unifying the API.

The test suite was expanded and refactored to test all these possibilities. The `a_structure.pdb` test PDB was slightly tweaked to avoid adding yet another file. I had to fix a number of other tests to accomodate these changes. All tests pass under Python 2.7 and Python 3.

Thanks @harryjubb for raising the problem that led to this fix, and @peterjc for the earlier work and usual leadership 👍 

I hereby agree to dual licence this and any previous contributions under both
the _Biopython License Agreement_ AND the _BSD 3-Clause License_.

I am happy to be thanked by name in the ``NEWS.rst`` and
``CONTRIB.rst`` files.

I have read the ``CONTRIBUTING.rst`` file and understand that AppVeyor and
TravisCI will be used to confirm the Biopython unit tests and ``flake8`` style
checks pass with these changes.

### CODE REVIEW

File: Bio/PDB/Residue.py [hunk 1]
```diff
@@ -54,19 +96,8 @@ def sort(self):
...
-            """
-            try:
-                i = ["N", "CA", "C", "O"].index(atom.name)
-            except ValueError:
-                i = 4
-            return (i, atom.name, atom.altloc)
-
-        self.child_list.sort(key=sort_index)
+
+        self.child_list.sort
```
Comment (peterjc) [inline Bio/PDB/Residue.py]: Should that not be ``self.child_list.sort()`` for an in-place sorting of the list?

**Comment by peterjc** [comment]
Is it worth adding public ``.sort()`` methods to each level of objects? i.e. Where there used to be a private ``._sort()`` method until #1189

File: Bio/PDB/Residue.py (hunk 1 above)
Comment (JoaoRodrigues) [inline Bio/PDB/Residue.py]: Well spotted!

**Comment by peterjc** [comment]
Are you happy with running ``flake8`` locally with the plugins as described in the CONTRIBUTING file? The TravisCI style checks are failing right now:

```
$ flake8 Bio/
Bio/PDB/Residue.py:92:1: D202 No blank lines allowed after function docstring
```

After that it will run ``flake8 Tests/`` as well.

**Comment by JoaoRodrigues** [comment]
> Is it worth adding public .sort() methods to each level of objects? i.e. Where there used to be a private ._sort() method until #1189

I would actually propose removing them altogether and have people use the `sorted` method instead. This is already supported anyway.

**Comment by peterjc** [comment]
OK - should we deprecate the single public ``.sort()`` method then?

For the git pre-commit hook, see issue #493 and possible implementation on #1188

File: Bio/PDB/Residue.py [hunk 2]
```diff
@@ -54,19 +96,8 @@ def sort(self):
...
-            The first integer is 0, 1, 2, 3, 4 for atoms N, CA, C, O, other.
-            """
-            try:
-                i = ["N", "CA", "C", "O"].index(atom.name)
-            except ValueError:
-                i = 4
-            return (i, atom.name, atom.altloc)
-
-        self.child_list.sort(key=sort_index)
+
```
Comment (peterjc) [inline Bio/PDB/Residue.py]: You'll need to remove this blank link to make ``flake8`` happy:

```
$ flake8 Bio/
Bio/PDB/Residue.py:92:1: D202 No blank lines allowed after function docstring
```

**Comment by JoaoRodrigues** [comment]
>OK - should we deprecate the single public .sort() method then?

I would vote for that.

**Comment by JoaoRodrigues** [comment]
@etal @peterjc Some tests are failing on Py 3.4 and made me realize I might have misunderstood how to write these methods. For example, for an `__eq__` method for equality, should I explicitly write what I expect to compare in the _other_ object - i.e. `return self.id == other.id` - or should I leave it to the language and implement it on both objects like `return self.id == other`? It seems that the latter is the best choice because it tries to figure out how to compare the objects first and if it cannot, then it raises the normal error 'Cannot compare objects x and y' instead of a missing attribute in case the _other_ object is say, a list.

File: Bio/PDB/Residue.py [hunk 3]
```diff
@@ -108,6 +108,10 @@ def sort(self):
         by name, with any alternative location specifier for disordered
         atoms (altloc) as a tie-breaker.
         """
+        warnings.warn("The custom sort() method will be deprecated in the "
```
Comment (peterjc) [inline Bio/PDB/Residue.py]: change "deprecated" to "removed" (or change the tense to "... has been deprecated in favour of...")

Comment (JoaoRodrigues) [inline Bio/PDB/Residue.py]: Done.

**Comment by JoaoRodrigues** [comment]
The coverage is decreasing because of all the rich comparison methods, which have a few conditionals in there, but the testing is actually more extensive than before. If others agree with these changes, we should merge this.

**Comment by peterjc** [comment]
Seems fine to me - @etal, @lennax, etc any comments before we merge this?

**Pull request merged**